fix: stuck in invalid state after last validation constraint is removed while invalid

Haprog opened the pull request:
Fixes #400

tomivirkki commented:
Quite a lot of changes in this PR. All the tests seem to pass with the following additional observer to `vaadin-text-field-mixin` (could be extended by numberfield if -mixin shouldn't know about min, max, step):

```
__constraintsChanged(required, minLength, maxLength, pattern, min, max, step) {
  if (!required && !minLength && !maxLength && !pattern && !min && !max) {
    this.invalid = false;
  }
  if (this.invalid) {
    this.validate();
  }
}
```
Am I missing something essential? If so, there should be more tests to cover all the changes.

Haprog commented:
> Quite a lot of changes in this PR. All the tests seem to pass with the following additional observer to `vaadin-text-field-mixin` (could be extended by numberfield if -mixin shouldn't know about min, max, step):
> 
> ```
> __constraintsChanged(required, minLength, maxLength, pattern, min, max, step) {
>   if (!required && !minLength && !maxLength && !pattern && !min && !max) {
>     this.invalid = false;
>   }
>   if (this.invalid) {
>     this.validate();
>   }
> }
> ```
> 
> Am I missing something essential? If so, there should be more tests to cover all the changes.

The problem with this simpler approach is that it will run validation only when all those are undefined, but not when there is still some active constraint but one or more other constraints were unset.

It's a shame that the multi dependency observers in Polymer don't provide the old values so you don't know which of those dependencies changed without manually tracking the old values elsewhere. The single property observer provides old value but I also ran to some issue when thinking about using individual observers.

I'll add a test case where there are many constraints applied and one is unset.

Haprog commented:
After taking a closer look, the simple `__constraintsChanged` observer approach seemed to work good after all considering all the cases I could think of.

I rebased this branch to latest master (with the flaky test fixes) and updated to use this simplified approach. Also added one more test and fixed some random typos in the test files.

tomivirkki commented on src/vaadin-text-field-mixin.html:
```diff
@@ -502,6 +503,18 @@
+        return;
+      }
+
+      if (!required && !minlength && !maxlength && !pattern && !min && !max) {
```
Oh, `min` and `max` can be set to 0 which is falsy value, but still a valid constraint. We should probably check for `undefined`/`null` here instead.

Haprog replied:
Good catch. I'm sure I was thinking about this but forgot it while changing other things.

Atm it makes an invalid field valid if e.g. `max` is set to `0` and value is bigger than `0`. I'll add test cases for these min/max 0 validity changes and fix this.

Haprog replied:
Fixing this revealed that the test `should update "invalid" state when "step" is removed` was a false positive before (it started failing). Fixed that test too and pushed an update.

Haprog replied:
Or well the test was correct, but was passing due to broken behaviour. Fixing that behaviour just revealed that another fix was needed too.
